File: src/tools/utility-scripts/commit_highlights.js
#!/usr/bin/env node
/**
 * Commit Highlights Tool for LLM-Generated Content
 *
 * This script removes <mark> tags from specified group IDs while preserving the content inside.
 * Designed for Task agents and LLMs to commit their approved highlights in walkthrough documents.
 *
 * Usage:
 *     node commit_highlights.js <file_path> <group_id>
 *     node commit_highlights.js <file_path> --all
 *
 * Examples:
 *     node commit_highlights.js interceptor-ts-deep-dive.md response-20250623103000
 *     node commit_highlights.js walkthrough-claude-trace-development.md --all
 */

import fs from "node:fs";

/**
 * Remove all <mark> tags for a specific group ID while preserving content.
 *
 * @param {string} content - File content as string
 * @param {string} groupId - The group ID to remove (e.g., "response-20250623103000")
 * @returns {string} Content with specified group's mark tags removed
 */
function removeGroupHighlights(content, groupId) {
	// Escape special regex characters in group ID
	const escapedGroupId = groupId.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");

	// Remove HTML comment delimiters for the group
	content = content.replace(
		new RegExp(`<!-- group-id:${escapedGroupId} -->\\n?`, "g"),
		"",
	);
	content = content.replace(
		new RegExp(`<!-- /group-id:${escapedGroupId} -->\\n?`, "g"),
		"",
	);

	// Remove mark tags with the specific group ID, preserving content
	const pattern = new RegExp(
		`<mark[^>]*data-group-id="${escapedGroupId}"[^>]*>(.*?)</mark>`,
		"gs",
	);
	content = content.replace(pattern, "$1");

	return content;
}

/**
 * Remove ALL <mark> tags while preserving content.
 *
 * @param {string} content - File content as string
 * @returns {string} Content with all mark tags removed
 */
function removeAllHighlights(content) {
	// Remove all HTML comment group delimiters
	content = content.replace(/<!-- group-id:[^>]+ -->\n?/g, "");
	content = content.replace(/<!-- \/group-id:[^>]+ -->\n?/g, "");

	// Remove all mark tags, preserving content
	content = content.replace(/<mark[^>]*>(.*?)<\/mark>/gs, "$1");

	return content;
}

/**
 * Find all group IDs in the content.
 *
 * @param {string} content - File content as string
 * @returns {string[]} List of group IDs found in the content
 */
function findGroupIds(content) {
	const pattern = /data-group-id="([^"]+)"/g;
	const matches = [];
	let match;

	match = pattern.exec(content);
	while (match !== null) {
		matches.push(match[1]);
		match = pattern.exec(content);
	}

	// Remove duplicates
	return [...new Set(matches)];
}

/**
 * Print usage information
 */
function printUsage() {
	console.log(`
Usage: node commit_highlights.js <file_path> [group_id|--all] [options]

Arguments:
  file_path               Path to the markdown file
  group_id               Group ID to commit (e.g., response-20250623103000)

Options:
  --all                  Remove all highlight marks
  --list                 List all group IDs in the file
  --dry-run              Show what would be changed without modifying file
  --help, -h             Show this help message

Examples:
  node commit_highlights.js interceptor-ts-deep-dive.md response-20250623103000
  node commit_highlights.js walkthrough-claude-trace-development.md --all
  node commit_highlights.js --list interceptor-ts-deep-dive.md
`);
}

function main() {
	const args = process.argv.slice(2);

	// Parse arguments
	if (args.length === 0 || args.includes("--help") || args.includes("-h")) {
		printUsage();
		process.exit(0);
	}

	const filePath = args[0];
	const groupId = args[1];
	const isAll = args.includes("--all");
	const isList = args.includes("--list");
	const isDryRun = args.includes("--dry-run");

	// Validate file path
	if (!fs.existsSync(filePath)) {
		console.error(`Error: File ${filePath} does not exist`);
		process.exit(1);
	}

	// Read file content
	let content;
	try {
		content = fs.readFileSync(filePath, "utf-8");
	} catch (error) {
		console.error(`Error reading file: ${error.message}`);
		process.exit(1);
	}

	// List group IDs if requested
	if (isList) {
		const groupIds = findGroupIds(content);
		if (groupIds.length > 0) {
			console.log("Found group IDs:");
			groupIds.sort().forEach((gid) => {
				console.log(`  - ${gid}`);
			});
		} else {
			console.log("No group IDs found in file");
		}
		return;
	}

	// Validate arguments
	if (!isAll && !groupId) {
		console.error("Error: Must specify either a group_id or --all");
		printUsage();
		process.exit(1);
	}

	if (isAll && groupId && !["--all", "--dry-run", "--list"].includes(groupId)) {
		console.error("Error: Cannot specify both group_id and --all");
		process.exit(1);
	}

	// Process content
	let newContent;
	if (isAll) {
		console.log(`Removing ALL highlight marks from ${filePath}`);
		newContent = removeAllHighlights(content);
	} else {
		console.log(`Removing highlights for group '${groupId}' from ${filePath}`);
		newContent = removeGroupHighlights(content, groupId);
	}

	// Check if any changes were made
	if (content === newContent) {
		if (isAll) {
			console.log("No highlight marks found to remove");
		} else {
			console.log(`No highlights found for group '${groupId}'`);
		}
		return;
	}

	// Show changes or write file
	if (isDryRun) {
		console.log("\n--- CHANGES (dry run) ---");
		// Simple diff indication
		const originalMarks = (content.match(/<mark[^>]*>/g) || []).length;
		const newMarks = (newContent.match(/<mark[^>]*>/g) || []).length;
		console.log(`Would remove ${originalMarks - newMarks} mark tags`);

		// Show sample of what would be removed
		if (groupId && !isAll) {
			const escapedGroupId = groupId.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
			const pattern = new RegExp(
				`<mark[^>]*data-group-id="${escapedGroupId}"[^>]*>`,
				"g",
			);
			const sampleMarks = content.match(pattern);
			if (sampleMarks && sampleMarks.length > 0) {
				const sample =
					sampleMarks[0].length > 100
						? `${sampleMarks[0].substring(0, 100)}...`
						: sampleMarks[0];
				console.log(`Sample mark to be removed: ${sample}`);
			}
		}
	} else {
		// Write the modified content back to file
		try {
			fs.writeFileSync(filePath, newContent, "utf-8");

			const originalMarks = (content.match(/<mark[^>]*>/g) || []).length;
			const newMarks = (newContent.match(/<mark[^>]*>/g) || []).length;
			console.log(
				`✅ Successfully removed ${originalMarks - newMarks} mark tags`,
			);
		} catch (error) {
			console.error(`Error writing file: ${error.message}`);
			process.exit(1);
		}
	}
}

// ES modules don't have require.main, so we check if this is the main module
if (import.meta.url === `file://${process.argv[1]}`) {
	main();
}

export { removeGroupHighlights, removeAllHighlights, findGroupIds };

File: src/tools/utility-scripts/commit_highlights.py
#!/usr/bin/env python3
"""Commit Highlights Tool for LLM-Generated Content

This script removes <mark> tags from specified group IDs while preserving the content inside.
Designed for Task agents and LLMs to commit their approved highlights in walkthrough documents.

Usage:
    python commit_highlights.py <file_path> <group_id>
    python commit_highlights.py <file_path> --all

Examples:
    python commit_highlights.py interceptor-ts-deep-dive.md response-20250623103000
    python commit_highlights.py walkthrough-claude-trace-development.md --all

"""

import argparse
import re
import sys
from pathlib import Path


def remove_group_highlights(content: str, group_id: str) -> str:
    """Remove all <mark> tags for a specific group ID while preserving content.
    
    Args:
        content: File content as string
        group_id: The group ID to remove (e.g., "response-20250623103000")
    
    Returns:
        Content with specified group's mark tags removed

    """
    # Remove HTML comment delimiters for the group
    content = re.sub(rf"<!-- group-id:{re.escape(group_id)} -->\n?", "", content)
    content = re.sub(rf"<!-- /group-id:{re.escape(group_id)} -->\n?", "", content)

    # Remove mark tags with the specific group ID, preserving content
    pattern = rf'<mark[^>]*data-group-id="{re.escape(group_id)}"[^>]*>(.*?)</mark>'
    content = re.sub(pattern, r"\1", content, flags=re.DOTALL)

    return content

def remove_all_highlights(content: str) -> str:
    """Remove ALL <mark> tags while preserving content.
    
    Args:
        content: File content as string
    
    Returns:
        Content with all mark tags removed

    """
    # Remove all HTML comment group delimiters
    content = re.sub(r"<!-- group-id:[^>]+ -->\n?", "", content)
    content = re.sub(r"<!-- /group-id:[^>]+ -->\n?", "", content)

    # Remove all mark tags, preserving content
    content = re.sub(r"<mark[^>]*>(.*?)</mark>", r"\1", content, flags=re.DOTALL)

    return content

def find_group_ids(content: str) -> list:
    """Find all group IDs in the content.
    
    Args:
        content: File content as string
    
    Returns:
        List of group IDs found in the content

    """
    pattern = r'data-group-id="([^"]+)"'
    matches = re.findall(pattern, content)
    return list(set(matches))  # Remove duplicates

def main():
    parser = argparse.ArgumentParser(
        description="Remove <mark> tags from LLM-generated content while preserving text",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  %(prog)s interceptor-ts-deep-dive.md response-20250623103000
  %(prog)s walkthrough-claude-trace-development.md --all
  %(prog)s --list interceptor-ts-deep-dive.md
        """
    )

    parser.add_argument("file_path", help="Path to the markdown file")
    parser.add_argument("group_id", nargs="?", help="Group ID to commit (e.g., response-20250623103000)")
    parser.add_argument("--all", action="store_true", help="Remove all highlight marks")
    parser.add_argument("--list", action="store_true", help="List all group IDs in the file")
    parser.add_argument("--dry-run", action="store_true", help="Show what would be changed without modifying file")

    args = parser.parse_args()

    # Validate file path
    file_path = Path(args.file_path)
    if not file_path.exists():
        print(f"Error: File {file_path} does not exist")
        sys.exit(1)

    # Read file content
    try:
        with open(file_path, encoding="utf-8") as f:
            content = f.read()
    except Exception as e:
        print(f"Error reading file: {e}")
        sys.exit(1)

    # List group IDs if requested
    if args.list:
        group_ids = find_group_ids(content)
        if group_ids:
            print("Found group IDs:")
            for gid in sorted(group_ids):
                print(f"  - {gid}")
        else:
            print("No group IDs found in file")
        return

    # Validate arguments
    if not args.all and not args.group_id:
        print("Error: Must specify either a group_id or --all")
        parser.print_help()
        sys.exit(1)

    if args.all and args.group_id:
        print("Error: Cannot specify both group_id and --all")
        sys.exit(1)

    # Process content
    if args.all:
        print(f"Removing ALL highlight marks from {file_path}")
        new_content = remove_all_highlights(content)
    else:
        print(f"Removing highlights for group '{args.group_id}' from {file_path}")
        new_content = remove_group_highlights(content, args.group_id)

    # Check if any changes were made
    if content == new_content:
        if args.all:
            print("No highlight marks found to remove")
        else:
            print(f"No highlights found for group '{args.group_id}'")
        return

    # Show changes or write file
    if args.dry_run:
        print("\n--- CHANGES (dry run) ---")
        # Simple diff indication
        original_marks = len(re.findall(r"<mark[^>]*>", content))
        new_marks = len(re.findall(r"<mark[^>]*>", new_content))
        print(f"Would remove {original_marks - new_marks} mark tags")

        # Show sample of what would be removed
        if args.group_id:
            pattern = rf'<mark[^>]*data-group-id="{re.escape(args.group_id)}"[^>]*>'
            sample_marks = re.findall(pattern, content)
            if sample_marks:
                print(f"Sample mark to be removed: {sample_marks[0][:100]}...")
    else:
        # Write the modified content back to file
        try:
            with open(file_path, "w", encoding="utf-8") as f:
                f.write(new_content)

            original_marks = len(re.findall(r"<mark[^>]*>", content))
            new_marks = len(re.findall(r"<mark[^>]*>", new_content))
            print(f"✅ Successfully removed {original_marks - new_marks} mark tags")

        except Exception as e:
            print(f"Error writing file: {e}")
            sys.exit(1)

if __name__ == "__main__":
    main()
